### PR TITLE
[session2_4] 김윤서 과제 제출합니다.

### PR DESCRIPTION
**Pull request by yseo17**
이번 딥러닝 개념은 지금까지 몇번 다루어봤던 내용이라 익숙했지만, 실습 부분에서는 약하다고 느껴 이번 기회에 mnist 모델 학습을 해볼 수 있었습니다. 개념 부분도 이전부터 연쇄 법칙과 역전파 부분을 이해하는 데에 특히 시간이 오래 걸렸던 기억이 있는데요, 이번 세션에서는 술 마시고 차인다는 비유가 재밌어서 축제 소리에도 재밌게 들을 수 있었습니다. 

과제는 실습을 복습하는 느낌으로 진행했는데, Adam 옵티마이저는 배워만 봤지 실제로 사용해본 건 이번이 처음이었네요! 과제 중에 성능 차이가 궁금해서 Adam 대신 SGD로도 한번 돌려보니 예측 정확도가 97%로 큰 차이가 나지 않더라고요.
이런 경우는 mnist 데이터셋이 비교적 단순하기 때문일까요? 또 더 크거나 복잡한 데이터/모델에서는 SGD는 수렴 속도가 느리거나 local minima에 빠질 가능성이 높아 Adam이 유리하다는 의미로 이해하면 될지 궁금합니다~

### CODE REVIEW

**Comment by dookongee** [comment]
과제 정말 잘하셨어요~~~

우선 질문하신 Adam과 SGD에 대해서는, 네 맞아요! SGD도 좋지만 local minima에 빠져서 더 이상 성능이 개선되지 않는 경우가 있지만, Adam은 그럴 가능성이 더 적어서  Adam이 유리한게 맞아요. 

질문하신 것에 대해서 추가로 더 말씀 드리자면...
현재 과제에서 진행한 mnist 데이터셋은 단순한 task에 해당합니다! 
일단 이미지 자체가 흑백이라서, 컴퓨터는 이미지에서 흰색이 있는지만 학습하면 돼요. 그리고 이미지 크기도 28x28이라서 매우 작죠. (하지만 보통 cnn으로 이미지 학습을 할 때는 이미지 크기가 256,256이나 448,448이 일반적이에요.)
 
즉, 28x28은 900칸 정도이고, 거기서 단순히 흰색이 어디에 있는지만 학습하면 되기 때문에 단순한 작업이라서, Loss 값도 3번 반복만에 0.05로 떨어졌어요.

하지만, 256x256은 65,536칸이라서 학습해야 할 값들 자체도 많죠.
 
그리고 지금은 흑백 이미지지만 만약 이게 컬러 이미지라면 더 복잡해져요. 
흑백 이미지는 gray scale 하나만 학습하지만, 컬러 이미지라면 r, g, b 3개를 학습해야 해요. 그래서 256x256 이미지라고 하면, 256x256 각 크기만큼의 r, g, b가 있어서 총 256x256x3 만큼을 학습하는 것과 같아요. 
숫자가 매우 크죠... 

추가로, 단순히 컬러 이미지가 아니라 face id와 같은 얼굴 인식 프로그램은 컴퓨터가 분류하기에 훨씬 더 복잡해요.

정리하면! mnist는 단순한 작업이기도 하고, 데이터의 크기도 매우 작아서 SGD나 Adam이나 차이가 없지만, 더 복잡한 상황에서는 일반적으로 Adam이 더 좋아요. 

그러면 항상 Adam이 좋냐? 그건 또 아니에요. 제가 11,12기때 강아지 감정 분석 모델을 만들 때, 성능이 좋은 모델 7개를 만들었는데, 그 중에서 5개는 Adam을 사용했지만, 2개는 SGD를 사용했어요. 대체적으로 Adam이 좋지만, SGD도 괜찮다는 것만 기억해주세요. 

Tip
1.모델 예측 성능이 90.1인데, 이것을 90.11 이렇게 조금이라도 올리는 것이 중요한 상황이라면 Adam말고 SGD를 사용했을 때 더 나은 결과가 나올 때도 있어요.
2. SGD는 Adam보다 초반에 loss 값을 더 빠르게 줄이는 특성이 있어요. 그래서 SGD를 사용해서 초반 epoch에서는 loss를 확확 줄이고, 후반 epoch은 모델 학습이 어느정도 진행된 상황이니까 후반에 가면 lr(learning rate)이 작아지도록 설정을 해서 loss를 천천히 줄이는 방법도 있어요. 
(좋은 질문을 해주셔서 신나게 떠들었네요 관련해서 궁금한거 있으면 언제든 물어보세요!!)

과제에서 꼭 기억하실 부분은 DataSet과 DataLodaer를 사용해서 학습한다는 것과 model.train(), model.eval()이에요. 
Dataset을 통해 데이터들을 이미지-정답 pair로 구성해놓고, 이것을 DataLoader를 사용해서 한 번에 batch_size라고 지정한 개수만큼의 데이터를 하나의 묶음으로 가져와요. 

model을 설계하고 나서 학습하기 위해서는 model.train()을 통해 학습모드로 설정해주고, 
학습된 모델로 테스트를 할 때는 model.eval()를 통해 평가모드로 설정해야 해요

cnn이 아니고 다른 어떤 딥러닝을 하더라도 공통적으로 사용하는 내용이니까 꼭 기억해주세요!!
